julia/src/native_function.jl

```diff
@@ -136,31 +133,35 @@ function NativeFunction(atom::Ptr{HsAtom}, ctx=PRELUDE, jit=JIT)
 end
 
 function (f::NativeFunction{R})(args...)::R where R
-    if length(args) != length(f.explicit_argument_signature)
-        throw(ArgumentError("wrong number of inputs, expected: $(length(f.explicit_argument_signature) )"))
-    end
-
     named_sizes = Dict{Symbol, Int}()
-    arg_ctypes = map(f.explicit_argument_signature, args) do binder, arg
-        to_ctype!(named_sizes, binder.type, arg)
+    num_explict = 0
+    c_args = map(f.argument_signature) do binder
+        if binder.implicit
+            -one(binder.type)  # we will do implicts in a second pass, once named_sizes is filled out
```

Comment:
Note that it is possible to handle the arguments in a single loop, but you'd have to iterate over them in reverse. The scoping rules for arguments are left-to-right, so if an implicit can be inferred from an argument, it has to come after it. So in the reverse loop it would get found first and would fill in the right fields in `named_sizes`.

Reply:
Ok cool!
lets add a note and then it can be changed later.
I will save that for a make that in a follow up PR, when I optimize this heavily.

```suggestion
            -one(binder.type)  # we will do implicts in a second pass, once named_sizes is filled out
            # Note: because scoping rules are left-to-right if we reversed the list we could do this in a single pass
```